Multiple simulations on the same GPU are interfering with each other!

If two or more simulations are running on the same GPU and using `TorchForce`, they produce incorrect results. Other GPU processes (i.e. VMD) also affect `TorchForce`, but the effects are less deterministic. Typically this manifests as random "explosions" of the system.

Versions:
- OpenMM-Torch is built from `master` using #9.
- PyTorch 1.4.0 is from https://anaconda.org/anaconda/pytorch

The problem can be reproduced with:
```python
from sys import stdout
from simtk import unit as u
import simtk.openmm as mm
from simtk.openmm import app

system = mm.System()
system.addParticle(1)

if True:
    import torch
    from openmmtorch import TorchForce

    class Force(torch.nn.Module):
        def forward(self, positions):
            return torch.sum(positions**2)
    torch.jit.script(Force()).save('model.pt')

    system.addForce(TorchForce('model.pt'))

else:
    force = mm.CustomExternalForce('x^2+y^2+z^2')
    force.addParticle(0)
    system.addForce(force)

integrator = mm.VerletIntegrator(1*u.femtosecond)
platform = mm.Platform.getPlatformByName('CUDA')
properties = {'CudaPrecision': 'mixed', 'DeviceIndex': '2'}

simulation = app.Simulation(app.Topology(), system, integrator, platform, properties)
simulation.context.setPositions([[1, 2, 3]])
simulation.reporters.append(app.StateDataReporter(stdout, 1000, step=True, totalEnergy=True, speed=True, separator='\t'))
simulation.step(100000)
```

If only one simulation is running, the total energy is conserved as expected:
```
#"Step" "Total Energy (kJ/mole)"        "Speed (ns/day)"
1000    14.000000179612432      0
2000    14.000005361777896      349
3000    14.000001366072164      367
4000    14.000005119669993      362
5000    14.000003331103173      336
6000    14.000002685719684      325
7000    14.000005601208187      328
8000    14.000000713319633      330
9000    14.000007006230264      338
10000   14.000000022274946      344
11000   14.00000683213267       349
12000   14.000000738332375      353
13000   14.00000512025951       350
14000   14.00000300080587       347
15000   14.00000395686458       332
16000   14.000004841749305      324
17000   14.000001304157916      318
18000   14.000006327369546      309
19000   14.000000236521004      303
20000   14.000007625364647      296
...
```
If two simulations are running, the energy conservation degrades:
```
#"Step" "Total Energy (kJ/mole)"        "Speed (ns/day)"
1000    14.000159708061787      0
2000    14.000443962602837      345
3000    14.000840991380645      335
4000    14.000969743602514      346
5000    14.001341178967788      344
6000    14.00135998002807       328
7000    14.001780500680749      313
8000    14.001861254934145      303
9000    14.001974845722069      297
10000   14.002129244844477      297
11000   14.002374428011779      299
12000   13.9968509309506        296
13000   14.020662965297788      291
14000   14.044255590957363      290
15000   14.058228920702625      291
16000   14.037448129767334      290
17000   14.030292677709525      283
18000   14.034618070407525      280
19000   14.037064627386965      278
20000   14.034805715731942      274
...
```
```
#"Step" "Total Energy (kJ/mole)"        "Speed (ns/day)"
1000    14.007740775328157      0
2000    14.015824312238987      260
3000    14.0087181449308        279
4000    14.022696349015083      284
5000    14.010502216207124      275
6000    14.012584191982722      265
7000    14.026440971902506      260
8000    14.030638631492987      260
9000    14.038191858222456      257
10000   14.043751931821303      252
11000   14.051324684105513      253
12000   14.055297832296048      252
13000   14.060540097186657      249
14000   14.08421743769247       250
15000   14.107282767704632      252
16000   14.101948425425924      253
17000   14.112720479074394      253
18000   14.130359066794343      252
19000   14.136352142606668      252
20000   14.146669114115653      252
...
```
For reference, I tried the same setup with `CustomExternalForce`. No problems have been observed!

I'm still struggling to get it to compile.  I switched to the conda build of PyTorch, but when I try to compile the plugin it gets pages of link errors like

```
/home/peastman/miniconda3/envs/pytorch/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so: undefined reference to `vdLn'
/home/peastman/miniconda3/envs/pytorch/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so: undefined reference to `vmsErf'
/home/peastman/miniconda3/envs/pytorch/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so: undefined reference to `zgetri_'
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -16,7 +16,7 @@ FIND_PACKAGE(Torch REQUIRED)
 #LINK_DIRECTORIES("${TENSORFLOW_DIR}/lib")
 
 # Specify the C++ version we are building for.
-SET (CMAKE_CXX_STANDARD 11)
+SET (CMAKE_CXX_STANDARD 14)
 
 # Set flags for linking on mac
 IF(APPLE)

--- a/platforms/cuda/src/CudaTorchKernels.cpp
+++ b/platforms/cuda/src/CudaTorchKernels.cpp
@@ -83,6 +83,7 @@ double CudaCalcTorchForceKernel::execute(ContextImpl& context, bool includeForce
     torch::Tensor energyTensor = module.forward(inputs).toTensor();
     if (includeForces) {
         energyTensor.backward();
+        // Note: "forceTensor" needs to be cloned due to a shared context (https://github.com/openmm/openmm-torch/issues/13)
         torch::Tensor forceTensor = posTensor.grad().clone();
         cu.setAsCurrent();
         void* data;

--- a/platforms/cuda/src/CudaTorchKernels.cpp
+++ b/platforms/cuda/src/CudaTorchKernels.cpp
@@ -83,7 +83,7 @@ double CudaCalcTorchForceKernel::execute(ContextImpl& context, bool includeForce
     torch::Tensor energyTensor = module.forward(inputs).toTensor();
     if (includeForces) {
         energyTensor.backward();
-        torch::Tensor forceTensor = posTensor.grad();
+        torch::Tensor forceTensor = posTensor.grad().clone();
         cu.setAsCurrent();
         void* data;
         if (cu.getUseDoublePrecision()) {

